documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,652 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+
+`Backends` is a mutable list of input backends. Until `Update` is called again, no device lists, state, etc on the context will be updated. The `ConnectionChanged` rules above will still be respected e.g. when you remove a backend, all of its devices will have a disconnected event raised for them.
+
+`Devices` contains all devices reported by all backends, including devices that do not necessarily fit into one of our more specialized wrapper lists. This means that if a backend has a device type we do not recognise, it will be accessible via this list.
+
+## Custom List Types
+
+These are relatively simple list wrappers with the events fired when state changes.
+
+```cs
+public partial class Mice : IReadOnlyList<IMouse>
+{
+    public event Action<ButtonEvent<IMouse, MouseButton>>? ButtonDown;
+    public event Action<ButtonEvent<IMouse, MouseButton>>? ButtonUp;
+    public event Action<ClickEvent<IMouse, MouseButton>>? Click;
+    public event Action<ClickEvent<IMouse, MouseButton>>? DoubleClick;
+    public event Action<AxisEvent<IMouse, Vector2>>? CursorMove;
+    public event Action<AxisEvent<IMouse, Vector2>>? Scroll;
+}
+
+public partial class Keyboards : IReadOnlyList<IKeyboard>
+{
+    public event Action<ButtonEvent<IKeyboard, Key>>? KeyDown;
+    public event Action<ButtonEvent<IKeyboard, Key>>? KeyUp;
+    public event Action<TextEvent<IKeyboard>>? TextInput;
+}
+
+public partial class Gamepads : IReadOnlyList<IGamepad>
+{
+    public event Action<ButtonEvent<IGamepad, JoystickButton>>? ButtonDown;
+    public event Action<ButtonEvent<IGamepad, JoystickButton>>? ButtonUp;
+    public event Action<AxisEvent<IGamepad, Vector2>>? ThumbstickMove;
+    public event Action<AxisEvent<IGamepad, float>>? TriggerMove;
+}
+
+public partial class Joysticks : IReadOnlyList<IJoystick>
+{
+    public event Action<ButtonEvent<IJoystick, JoystickButton>>? ButtonDown;
+    public event Action<ButtonEvent<IJoystick, JoystickButton>>? ButtonUp;
+    public event Action<AxisEvent<IJoystick, float>>? AxisMove;
+    public event Action<AxisEvent<IJoystick, Vector2>>? HatMove;
+}
+```
+
+Unlike 1.0 and 2.0, this proposal uses `readonly record struct`s as their only argument for the event action. This allows us to provide more information to the event handlers without breaking in the future. These types are farily simple:
+
+```cs
+public readonly record struct ButtonEvent<TDevice, TButton>(TDevice DeviceState, TButton Button);
+public readonly record struct ClickEvent<TDevice, TButton>(TDevice DeviceState, TButton Button, Vector2 Position);
+public readonly record struct AxisEvent<TDevice, TAxis>(TDevice DeviceState, int Index, TAxis OldValue, TAxis NewValue);
+public readonly record struct TextEvent<TDevice>(TDevice DeviceState, string? OldText, string? NewText);
```

Comment:
```suggestion
public readonly record struct ButtonEvent<TDevice, TButton>(TDevice sender, TButton button);
public readonly record struct ClickEvent<TDevice, TButton>(TDevice sender, TButton button, Vector2 position);
public readonly record struct AxisEvent<TDevice, TAxis>(TDevice sender, int index, TAxis oldValue, TAxis newValue);
public readonly record struct TextEvent<TDevice>(TDevice sender, string? oldText, string? newText);
```

> Yeah, but with all of the events you have a "sender" as the first argument telling you which IWhatever caused the event to happen
Then these arguments should probably not be "state", but rather the device itself (also, argument names should start with lowercase)

Reply:
> Then these arguments should probably not be "state"

Good catch, left over from a previous revision.

> (also, argument names should start with lowercase)

Not for records, these map into the names of the properties/fields so unless you want to adopt Unity's nomenclature - no thanks!

```suggestion
public readonly record struct ButtonEvent<TDevice, TButton>(TDevice Device, TButton Button);
public readonly record struct ClickEvent<TDevice, TButton>(TDevice Device, TButton Button, Vector2 Position);
public readonly record struct AxisEvent<TDevice, TAxis>(TDevice Device, int Index, TAxis OldValue, TAxis NewValue);
public readonly record struct TextEvent<TDevice>(TDevice Device, string? OldText, string? NewText);
```